batched/dense/impl/KokkosBatched_Pttrs_Serial_Internal.hpp

```diff
@@ -0,0 +1,138 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+#ifndef KOKKOSBATCHED_PTTRS_SERIAL_INTERNAL_HPP_
+#define KOKKOSBATCHED_PTTRS_SERIAL_INTERNAL_HPP_
+
+#include <KokkosBatched_Util.hpp>
+
+/// \author Yuuichi Asahi (yuuichi.asahi@cea.fr)
+
+namespace KokkosBatched {
+
+template <typename ArgUplo, typename AlgoType>
+struct SerialPttrsInternal {
+  template <typename ValueType>
+  KOKKOS_INLINE_FUNCTION static int invoke(
+      const int n, const ValueType *KOKKOS_RESTRICT d, const int ds0,
+      const ValueType *KOKKOS_RESTRICT e, const int es0,
+      ValueType *KOKKOS_RESTRICT b, const int bs0, const int ldb);
+
+  template <typename ValueType>
+  KOKKOS_INLINE_FUNCTION static int invoke(
+      const int n, const ValueType *KOKKOS_RESTRICT d, const int ds0,
+      const Kokkos::complex<ValueType> *KOKKOS_RESTRICT e, const int es0,
+      Kokkos::complex<ValueType> *KOKKOS_RESTRICT b, const int bs0,
+      const int ldb);
+};
+
+///
+/// Real matrix
+///
+
+template <>
+template <typename ValueType>
+KOKKOS_INLINE_FUNCTION int
+SerialPttrsInternal<Uplo::Lower, Algo::Pttrs::Unblocked>::invoke(
+    const int n, const ValueType *KOKKOS_RESTRICT d, const int ds0,
+    const ValueType *KOKKOS_RESTRICT e, const int es0,
+    ValueType *KOKKOS_RESTRICT b, const int bs0, const int ldb) {
+  // Solve A * X = B using the factorization L * D * L**T
+  for (int i = 1; i < n; i++) {
+    b[i * bs0] -= e[(i - 1) * es0] * b[(i - 1) * bs0];
+  }
+
+  b[(n - 1) * bs0] /= d[(n - 1) * ds0];
+
+  for (int i = n - 2; i >= 0; i--) {
+    b[i * bs0] = b[i * bs0] / d[i * ds0] - b[(i + 1) * bs0] * e[i * es0];
+  }
+
+  return 0;
+}
+
+template <>
+template <typename ValueType>
+KOKKOS_INLINE_FUNCTION int
+SerialPttrsInternal<Uplo::Upper, Algo::Pttrs::Unblocked>::invoke(
```

Comment:
Since the matrix is symmetric `Uplo::Lower` and `Uplo::Upper` end up being the same implementation, maybe we can just put that in a single function and call it in both specializations? That's less code maintenance and potential bugs in the future?